internal/cri/server/images/image_pull.go

```diff
@@ -180,51 +178,91 @@ func (c *CRIImageService) PullImage(ctx context.Context, name string, credential
-	//pullOpts = append(pullOpts, c.encryptedImagesPullOpts()...)
-	if !c.config.DisableSnapshotAnnotations {
-		pullOpts = append(pullOpts,
-			containerd.WithImageHandlerWrapper(snpkg.AppendInfoHandlerWrapper(ref)))
-	}
+		log.G(ctx).Debugf("PullImage %q with snapshotter %s using client.Pull()", ref, snapshotter)
+		pullOpts := []containerd.RemoteOpt{
+			containerd.WithSchema1Conversion, //nolint:staticcheck // Ignore SA1019. Need to keep deprecated package for compatibility.
+			containerd.WithResolver(resolver),
+			containerd.WithPullSnapshotter(snapshotter),
+			containerd.WithPullUnpack,
+			containerd.WithPullLabels(labels),
+			containerd.WithMaxConcurrentDownloads(c.config.MaxConcurrentDownloads),
+			containerd.WithImageHandler(imageHandler),
+			containerd.WithUnpackOpts([]containerd.UnpackOpt{
+				containerd.WithUnpackDuplicationSuppressor(c.unpackDuplicationSuppressor),
+				containerd.WithUnpackApplyOpts(diff.WithSyncFs(c.config.ImagePullWithSyncFs)),
+			}),
+		}
 
-	if c.config.DiscardUnpackedLayers {
-		// Allows GC to clean layers up from the content store after unpacking
-		pullOpts = append(pullOpts,
-			containerd.WithChildLabelMap(containerdimages.ChildGCLabelsFilterLayers))
-	}
+		// Temporarily removed for v2 upgrade
+		//pullOpts = append(pullOpts, c.encryptedImagesPullOpts()...)
+		if !c.config.DisableSnapshotAnnotations {
+			pullOpts = append(pullOpts,
+				containerd.WithImageHandlerWrapper(snpkg.AppendInfoHandlerWrapper(ref)))
+		}
 
-	pullReporter.start(pctx)
-	image, err := c.client.Pull(pctx, ref, pullOpts...)
-	pcancel()
-	if err != nil {
-		return "", fmt.Errorf("failed to pull and unpack image %q: %w", ref, err)
+		if c.config.DiscardUnpackedLayers {
+			// Allows GC to clean layers up from the content store after unpacking
+			pullOpts = append(pullOpts,
+				containerd.WithChildLabelMap(containerdimages.ChildGCLabelsFilterLayers))
+		}
+
+		pullReporter.start(pctx)
+		image, err = c.client.Pull(pctx, ref, pullOpts...)
+		pcancel()
+		if err != nil {
+			return "", fmt.Errorf("failed to pull and unpack image %q: %w", ref, err)
+		}
+	} else { // Use transfer service by default
+		log.G(ctx).Debugf("PullImage %q with snapshotter %s using transfer service", ref, snapshotter)
+		rctx, rcancel := context.WithCancel(ctx)
+		defer rcancel()
+		transferProgressReporter := newTransferProgressReporter(ref, rcancel, imagePullProgressTimeout)
+
+		// Set image store opts
+		var sopts []transferimage.StoreOpt
+		sopts = append(sopts, transferimage.WithPlatforms(platforms.DefaultSpec()))
+		sopts = append(sopts, transferimage.WithUnpack(platforms.DefaultSpec(), snapshotter))
+		sopts = append(sopts, transferimage.WithImageLabels(labels))
+		is := transferimage.NewStore(ref, sopts...)
+		log.G(ctx).Debugf("Getting new CRI credentials")
+		ch := newcriCredentials(ref, credentials)
+		opts := []registry.Opt{registry.WithCredentials(ch)}
+		reg, _ := registry.NewOCIRegistry(ctx, ref, opts...)
```

Comment:
Any reason to ignore the returned err here?

Reply:
Returning error now.